example-crates/example_esplora/src/main.rs

```diff
@@ -0,0 +1,326 @@
+                            index
+                        );
+                        script
+                    })));
+                }
+                if utxos {
+                    // We want to search for whether the UTXO is spent, and spent by which
+                    // transaction. We provide the outpoint of the UTXO to
+                    // `EsploraExt::update_tx_graph_without_keychain`.
+                    let init_outpoints = graph.index.outpoints().iter().cloned();
+                    let utxos = graph
+                        .graph()
+                        .filter_chain_unspents(&*chain, chain_tip, init_outpoints)
+                        .map(|(_, utxo)| utxo)
+                        .collect::<Vec<_>>();
+                    outpoints = Box::new(
+                        utxos
+                            .into_iter()
+                            .inspect(|utxo| {
+                                eprintln!(
+                                    "Checking if outpoint {} (value: {}) has been spent",
+                                    utxo.outpoint, utxo.txout.value
+                                );
+                            })
+                            .map(|utxo| utxo.outpoint),
+                    );
+                };
+                if unconfirmed {
+                    // We want to search for whether the unconfirmed transaction is now confirmed.
+                    // We provide the unconfirmed txids to
+                    // `EsploraExt::update_tx_graph_without_keychain`.
+                    let unconfirmed_txids = graph
+                        .graph()
+                        .list_chain_txs(&*chain, chain_tip)
+                        .filter(|canonical_tx| !canonical_tx.observed_as.is_confirmed())
+                        .map(|canonical_tx| canonical_tx.node.txid)
+                        .collect::<Vec<Txid>>();
+                    txids = Box::new(unconfirmed_txids.into_iter().inspect(|txid| {
+                        eprintln!("Checking if {} is confirmed yet", txid);
+                    }));
+                }
+            }
+
+            client.update_tx_graph_without_keychain(
+                spks,
+                txids,
+                outpoints,
+                scan_options.parallel_requests,
+            )?
+        }
+    };
+
+    println!();
+
+    // Up to this point, we have only created a `TxGraph` update, but not an update for our
+    // `ChainOracle` implementation (`LocalChain`). The `TxGraph` update may contain chain anchors,
+    // so we need the corresponding blocks to exist in our `LocalChain`. Here, we find the heights
+    // of missing blocks in `LocalChain`.
+    //
+    // Getting the local chain tip is only for printing to stdout.
+    let (missing_block_heights, tip) = {
+        let chain = &*chain.lock().unwrap();
+        let heights_to_fetch = graph_update.missing_heights(chain).collect::<Vec<_>>();
```

Comment:
I think this can be simplified. Notice we can figure out heights_to_fetch by just iterating over the TxGraph changeset's added anchors. Each new anchor height is almost certainly relevant to us and can't see any harm in just fetching the block at that height even if somehow it's not. I'm glad that `missing_heights` exists and is correct but I don't think it's needed here.

This means this example can be changed the update is applied and an `indexed_tx_graph::ChangeSet` is returned from each match arm. Then we can go and update the chain. We *could* even persist the change in each match arm and lazily update the chain and persist that afterwards as this demonstrates the decoupling of chain and graph nicely.

Reply:
Unresolving this since it was only partially addressed. Here's what I think is remaining to do:

Do not declare a mutable changeset outside the match arms and the progressively update it. Instead just make each match arm return a `indexed_tx_graph::ChangeSet`. This means the graph part of the update is completed in the match arms and then you only have to do the local chain update outside. This is feels simpler and makes the separation between the two updates more stark.